### PR TITLE
Changes to fix while reading BigLake table in connector. 

### PR DESCRIPTION
**Pull request by suryasoma**
Changes to fix while reading BigLake table in connector. Added integration tests as well

### CODE REVIEW

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java [hunk 1]
```diff
@@ -68,6 +72,25 @@ public static void runQuery(String query) {
     bigQueryClient.query(query);
   }
 
+  public static void createExternalTable(
+      String dataset,
+      String table,
+      StructType schemaStruct,
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java:78]: nit: s/schemaStruct/schema

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 2]
```diff
@@ -67,7 +68,23 @@ public class TestConstants {
   static final DecimalType NUMERIC_SPARK_TYPE =
       DataTypes.createDecimalType(BQ_NUMERIC_PRECISION, BQ_NUMERIC_SCALE);
   static final String SHAKESPEARE_TABLE = "bigquery-public-data.samples.shakespeare";
+  static final String BIGLAKE_SHAKESPEARE_TABLE = "spark_bq_connector_dataset.biglake-shakespeare";
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:71]: remove constant table names

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 3]
```diff
@@ -67,7 +68,23 @@ public class TestConstants {
...
   static final long SHAKESPEARE_TABLE_NUM_ROWS = 164656L;
+  static final String TEMPORARY_GCS_BUCKET_ENV_VARIABLE = "TEMPORARY_GCS_BUCKET";
+  static final String BIGLAKE_CONNECTION_ID_ENV_VARIABLE = "BIGLAKE_CONNECTION_ID";
+  static final String temporaryGcsBucket =
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:75]: const should adhere to the name standard: `TEMPORARY_GCS_BUCKET`

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java [hunk 4]
```diff
@@ -67,7 +68,23 @@ public class TestConstants {
...
+      Preconditions.checkNotNull(
+          System.getenv(TEMPORARY_GCS_BUCKET_ENV_VARIABLE),
+          "Please set the %s env variable to point to a write enabled GCS bucket",
+          TEMPORARY_GCS_BUCKET_ENV_VARIABLE);
+  static final String bigLakeConnectionID =
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java:80]: const should adhere to the name standard: `IGLAKE_CONNECTION_ID`

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 5]
```diff
@@ -375,4 +379,112 @@ public void testArrowCompressionCodec() {
...
+          destinationURI,
+          contentType);
+    } catch (Exception exception) {
+      exception.printStackTrace();
+    }
+  }
+
+  @Test
+  public void testReadFromBigLakeTable_csv() {
+    String temporaryGCSURI =
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:396]: style: s/temporaryGCSURI/temporaryGcsUri

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 6]
```diff
@@ -375,4 +379,112 @@ public void testArrowCompressionCodec() {
...
+          getClass().getResourceAsStream("/integration/" + resourceName),
+          destinationURI,
+          contentType);
+    } catch (Exception exception) {
+      exception.printStackTrace();
+    }
+  }
+
+  @Test
+  public void testReadFromBigLakeTable_csv() {
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:395]: All ` testReadFromBigLakeTable_*` tests can use the same internal implementation with different parameters. No need for code redundancies

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 7]
```diff
@@ -375,4 +379,112 @@ public void testArrowCompressionCodec() {
...
     assertThat(avroResults).isEqualTo(arrowResultsForLZ4FrameCodec);
   }
+
+  private void uploadFileToGCS(String resourceName, String destinationURI, String contentType) {
+    try {
+      AcceptanceTestUtils.uploadToGcs(
+          getClass().getResourceAsStream("/integration/" + resourceName),
+          destinationURI,
+          contentType);
+    } catch (Exception exception) {
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:389]: style: s/Exception exception/Exception e

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 8]
```diff
@@ -375,4 +379,112 @@ public void testArrowCompressionCodec() {
...
+      exception.printStackTrace();
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:390]: Never call `printStackTrace()`. Instead throw a new RuntimeException wrapping the original one with a comment

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java [hunk 9]
```diff
@@ -375,4 +379,112 @@ public void testArrowCompressionCodec() {
...
+            .format("bigquery")
+            .option("dataset", testDataset.toString())
+            .option("table", table)
+            .load();
+    testShakespeare(df);
+  }
+
+  @Test
+  public void testReadFromBigLakeTable_avro() {
+    String temporaryGCSURI =
```
Comment (davidrabinowitz) [inline spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java:443]: this method should have assertThrows(), as avro is not supported yet

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java (hunk 5 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java (hunk 4 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java (hunk 3 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java (hunk 2 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java (hunk 1 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java (hunk 4 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/TestConstants.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java (hunk 8 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java (hunk 9 above)
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/ReadIntegrationTestBase.java]: Addressed

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 10]
```diff
@@ -378,31 +378,34 @@ public long calculateTableSize(TableId tableId, Optional<String> filter) {
...
-      throw new BigQueryConnectorException(
-          "Querying table size was interrupted on the client side", e);
+    TableDefinition.Type type = tableInfo.getDefinition().getType();
+    if (type == TableDefinition.Type.TABLE && !filter.isPresent()) {
+      return tableInfo.getNumRows().longValue();
+    } else if (type == TableDefinition.Type.EXTERNAL && !filter.isPresent()) {
+      String table = fullTableName(tableInfo.getTableId());
+      return getNumberOfRows(String.format("SELECT COUNT(*) from `%s`", table));
+    } else if (type == TableDefinition.Type.VIEW
+        || type == TableDefinition.Type.MATERIALIZED_VIEW
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:388]: @sonatype-lift ignore

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 11]
```diff
@@ -378,31 +378,34 @@ public long calculateTableSize(TableId tableId, Optional<String> filter) {
...
+        || (type == TableDefinition.Type.TABLE && filter.isPresent())) {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:389]: @sonatype-lift ignore

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 12]
```diff
@@ -378,31 +378,34 @@ public long calculateTableSize(TableId tableId, Optional<String> filter) {
...
-    } catch (InterruptedException e) {
-      throw new BigQueryConnectorException(
-          "Querying table size was interrupted on the client side", e);
+    TableDefinition.Type type = tableInfo.getDefinition().getType();
+    if (type == TableDefinition.Type.TABLE && !filter.isPresent()) {
+      return tableInfo.getNumRows().longValue();
+    } else if (type == TableDefinition.Type.EXTERNAL && !filter.isPresent()) {
+      String table = fullTableName(tableInfo.getTableId());
+      return getNumberOfRows(String.format("SELECT COUNT(*) from `%s`", table));
+    } else if (type == TableDefinition.Type.VIEW
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java:387]: @sonatype-lift ignore

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (suryasoma) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java]: Addressed

File: bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java [hunk 13]
```diff
@@ -378,31 +378,34 @@ public long calculateTableSize(TableId tableId, Optional<String> filter) {
...
+    } else {
+      throw new IllegalArgumentException(
+          String.format(
+              "Unsupported table type %s for table %s",
+              type, fullTableName(tableInfo.getTableId())));
     }
   }
 
+  private long getNumberOfRows(String sql) {
+    long numberOfRows;
```
[the remark this answers is not in the document]
Comment (suryasoma) [reply bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryClient.java]: Addressed

File: spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java [hunk 14]
```diff
@@ -68,6 +71,25 @@ public static void runQuery(String query) {
     bigQueryClient.query(query);
   }
 
+  public static void createExternalTable(
```
[the remark this answers is not in the document]
Comment (suryasoma) [reply spark-bigquery-tests/src/main/java/com/google/cloud/spark/bigquery/integration/IntegrationTestUtils.java]: Addressed

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**